# Scripts for the Low coverage DFTD project


Preliminary list of codes that need to be placed:

In this repository, you will find the pipeline scripts for the following:

In the Mitochondrial Analysis folder
1) Calling and Filtering Variants in the Mitochondria --added
2) Quantification of Mitochondrial Copy Number --added (may need to delete this, since it's not in paper?)
3) Phylogeny using Mitochondrial Variants --added
4) Plot of Spread of Mitochondrial haplotypes across Tasmania --combine with the nuclear genome maps.
5) Should we include the Forestier and West Pencil Pine analysis? If at all, it should be in the supplementary?

In the Nuclear Genome Analysis folder

6) Calling and filtering of variants from the nuclear genome--added but will need to join into single variant calling pipeline script
7) Calling copy numbers on low-coverage samples
8) Generating Phylogeny using SNVs from the Nuclear Genome + somatic mitochondrial variants 
9) QC and filtering the initial 600 probe panel (MIPs)
10) Filtering of MIPs (post-QC) and overlapping with copy number states
11) Generating Phylogeny using MIPs
12) Telomere analysis
13) Plotting spread of different groups across Tasmania
