# Scripts for the Low coverage DFTD project


Preliminary list of codes that need to be placed:

In this repository, you will find the pipeline scripts for the following:
1) Calling and Filtering Variants in the Mitochondria
2) Quantification of Mitochondrial Copy Number
3) Phylogeny using Mitochondrial Variants
4) Plot of Spread of Mitochondrial haplotypes across Tasmania

5) Calling and filtering of variants from the nuclear genome
6) Calling copy numbers on low-coverage samples 
7) Generating Phylogeny using SNVs from the Nuclear Genome + somatic mitochondrial variants
8) QC and filtering the initial 600 probe panel (MIPs)
9) Filtering of MIPs (post-QC) and overlapping with copy number states
10) Generating Phylogeny using MIPs
11) Telomere analysis

12) Plotting spread of different groups across Tasmania
